feat(config)!: move to using PaasConfig CRD instead of yaml file for config

hikarukin opened the pull request:
This PR migrates Paas Operator configuration from using a yaml file to a CRD.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [x] Other (please describe): tests

## What is the current behavior?

Currently, operator config is stored in a yaml file / ConfigMap.

## What is the new behavior?

This PR migrates Paas Operator configuration from using a yaml file to a CRD. The operator will no longer be using CM but CRD instead. It will also auto reload config when CRD is updated.

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

Adminstrators of the operator will need to migrate their configuration from yaml file / CM to CRD. No config options were removed. Some config options were made optional by the CRD providing default values.

## Other information

Some verification behaviour is migrated from custom Verify functions into the CRD spec using +kubebuilder:validation markers.

hikarukin committed:
delete unused config.go

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
delete unused config.go

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add paasconfig api type and generated CRD

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add paasconfig api type and generated CRD

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
adjust for added api and spec

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
adjust for added api and spec

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
change ratio to int64 to prevent unsafe type in crd

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
change ratio to int64 to prevent unsafe type in crd

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
correct generated manifest for ratio adjustment

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
correct generated manifest for ratio adjustment

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
update copyright year in boilerplate

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
update copyright year in boilerplate

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 commented on manifests/crds/cpet.belastingdienst.nl_paasconfig.yaml:
Please add to kustomization.yaml in the directory

hikarukin committed:
add paasconfig to kustomization.yaml

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add paasconfig to kustomization.yaml

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
Updates markers for paasconfig (#170)

* * Updates PROJECT for PaasConfig

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

* * Adds kubebuilder markers for PaasConfig

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

---------

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
Updates markers for paasconfig (#170)

* * Updates PROJECT for PaasConfig

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

* * Adds kubebuilder markers for PaasConfig

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

---------

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Description for type paasconfig_controller.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Description for type paasconfig_controller.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Workaround make generate by using custom namespacedName

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Workaround make generate by using custom namespacedName

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Removes unused `retries` Argopermissions config (#172)

* Adds descriptions in paasconfig_types.go
* Finetunes required/optional in paasconfig_types.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Removes unused `retries` Argopermissions config (#172)

* Adds descriptions in paasconfig_types.go
* Finetunes required/optional in paasconfig_types.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Rebase on upstream main
* Make lint-fix

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
Merge branch 'feat-add-paasconfig-api' into rebase-config-api

# Conflicts:
#	api/v1alpha1/paasconfig_types.go
#	api/v1alpha1/zz_generated.deepcopy.go
#	cmd/manager/main.go
#	internal/controller/argocd.go
#	internal/controller/capabilities.go
#	internal/controller/cluster_quotas.go
#	internal/controller/clusterrolebindings.go
#	internal/controller/ldap_groups.go
#	internal/controller/main.go
#	internal/controller/namespaces.go
#	manifests/config/cm_opr-paas.yaml
#	manifests/crds/cpet.belastingdienst.nl_paasconfig.yaml

hikarukin committed:
Call Verify on PaasConfig

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Prevent retry of reconcile

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Add defaultPolicy to PaasConfig test fixture

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Finalize PaasConfig and start adding a Status Condition to PaasConfig

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
add PaasConfig example instead of ConfigMap

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
remove unneeded fixture

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
cleanup

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
change dep to indirect

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Merge branch 'main' into feat-add-paasconfig-api

portly-halicore-76 commented on manifests/config/kustomization.yaml:
```diff
@@ -3,5 +3,4 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
 resources:
-  - cm_opr-paas.yaml
```
Add sample paasconfig here, else we don't provide a complete default installation as used by e2e-tests against deployed operator. If we would like to do this in another way, we must describe and alter CI which is also fine by me. For example by really providing an example config in de examples directory and using that in CI...

hikarukin replied:
For now I copied the existing example from the examples directory and added it to kustomization.yaml

hikarukin committed:
add docs

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
rename docs file

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add quick example paasconfig

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin commented:
I added a bunch of changes to docs, not all related to this PR directly but they impact the way the docs are shown.

Related to this PR I added an auto-generated API reference and some tips on writing proper docs. My intent is that we run the generation of the API reference through an action at a later date.

portly-halicore-76 committed:
* Adds appropriate status somewhere

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
Merge remote-tracking branch 'upstream/main' into feat-add-paasconfig-api

# Conflicts:
#	docs/development-guide/e2e-test-details.md
#	manifests/config/cm_opr-paas.yaml
#	test/e2e/fixtures/paas_config.yml

portly-halicore-76 committed:
* Merge

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Cleanup

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Aligns test-e2e against deployed image config with test-e2e config.
* Adds some todo's which can break the operator on restarts

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Adds more complex determination of singleton pattern.

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Adds more complex determination of singleton pattern. Is this OK?

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* What to do with the last required object in the Verify() method, seems not doing what it should do

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Just store PaasConfigSpec as currentConfig instead of whole PaasConfig.

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Adds empty config check in Paas(Ns) reconcilers

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
Replace start-e2e target by run-operator target

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Removes verify entirely
* Clean-up TODO's
* Add message to paas(ns) when no config is found

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
fix minor linter issues

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add initial PaasConfig test

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Merge branch 'main' into feat-add-paasconfig-api

portly-halicore-76 committed:
* Make lint-fix
* Updates deprecations

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
fix test

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 commented on test/e2e/paasconfig_test.go:
```diff
@@ -0,0 +1,40 @@
+
+	expectedStatus := v1.ConditionStatus("True")
+	condition := fetchedPaasConfig.Status.Conditions[0]
+	assert.Equal(t, expectedStatus, condition.Status)
```
assert.True(t, meta.IsStatusConditionPresentAndEqual(fetchedPaasConfig.Status.Conditions, v1alpha1.TypeActivePaasConfig, metav1.ConditionTrue))

hikarukin committed:
cleanup test and add config is updated test

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
add assertPaasConfigInvalidSpec

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
update manifests

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
go mod tidy

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
fix e2e testing

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
fix broken deploy

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
apply crds first

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
wip

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 commented on Makefile:
```diff
@@ -225,6 +213,7 @@ uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified
 .PHONY: deploy
 deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
 	cd manifests/manager && $(KUSTOMIZE) edit set image controller=${IMG}
+	$(KUSTOMIZE) build manifests/crds | $(KUBECTL) apply -f -
```
This is redundant as they are part of manifests/default..

hikarukin replied:
It is added to resolve a timing issue between CRD creation and PaasConfig resource instance creation. Locally it was going fine, but on actions there appeared to be a resource issue. PaasConfig CRD wasn't available yet when it was already trying to create the resource. This resolved it, though nasty.

hikarukin committed:
fix CICD by using env var to control secrets path

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Fixes e2e-tests

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

hikarukin committed:
Revert "fix CICD by using env var to control secrets path"

This reverts commit 3c7f3209f30b1bbfbfbb44b0cae28569127d8503.

hikarukin committed:
move example config to examples dir

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

portly-halicore-76 committed:
* Fixes https://github.com/belastingdienst/opr-paas/issues/181 (#183)

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>
Co-authored-by: Hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
update docs to reflect having to apply config

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Fix tests

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
changed case

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
fix makefile

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
fix teardown registration

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>

hikarukin committed:
Merge branch 'main' into feat-add-paasconfig-api

portly-halicore-76 committed:
* Adds deprecation for decryptSecrets
* Minor cleanup redundant code main_test.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 committed:
* Adds deprecation for decryptSecrets
* Minor cleanup redundant code main_test.go

Signed-off-by: portly-halicore-76 <170707699+portly-halicore-76@users.noreply.github.com>

portly-halicore-76 reviewed (dismissed):
What a nice improvement.

hikarukin committed:
small cleanup

Signed-off-by: hikarukin <156204784+hikarukin@users.noreply.github.com>
